### PR TITLE
implement: mclock, update_rtm, update_stablevv

### PR DESCRIPTION
**Pull request by gyounes**

### CODE REVIEW

File: include/ishikawa.hrl [hunk 1]
```diff
@@ -1,3 +1,10 @@
 -type actor() :: binary().
 -type message() :: binary().
 -type timestamp() :: vclock:vclock().
+-type timestampMatrix() :: mclock:mclock().
```
Comment (cmeiklejohn) [inline include/ishikawa.hrl]: `timestamp_matrix` would be more idiomatic: the only thing that's usually CamelCase in Erlang are variables.

File: include/ishikawa.hrl [hunk 2]
```diff
@@ -1,3 +1,10 @@
...
+-record(state, {actor :: actor(),
+                vv :: timestamp(),
+                members :: [term()],
+                svv :: timestamp(),
+                rtm :: timestampMatrix(),
+                toBeDlvrdQ :: [{timestamp(), message()}]}).
```
Comment (cmeiklejohn) [inline include/ishikawa.hrl]: It would be useful to switch this to snake_case and be more explicit about the naming -- I don't easily know what this is.

File: src/ishikawa_backend.erl [hunk 3]
```diff
@@ -45,10 +45,6 @@
 
 -define(PEER_SERVICE, partisan_peer_service).
 
--record(state, {actor :: actor(),
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl:48]: Since each module may have it's own state, the state should stay here and not be in an include file.

File: src/ishikawa_backend.erl [hunk 4]
```diff
@@ -126,9 +131,19 @@ handle_call({tcbcast, Message}, _From, #state{actor=Actor,
...
+                                              rtm=RTM0,
+                                              toBeDlvrdQ=Queue0} = State) ->
+    %% Check if the message should be delivered
+    {VClock, Queue} = trcb:check_causal_delivery({Message, Timestamp}, From, VClock0, Queue0),
+
+    %% Update the Recent Timestamp Matrix
+    RTM = mclock:update_rtm(RTM0, From, Timestamp),
+
+    %% Update the Stable Version Vector
+    SVV = mclock:update_stablevv(RTM0),
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: Shouldn't this be `RTM` and not `RTM0`?

File: src/mclock.erl [hunk 5]
```diff
@@ -0,0 +1,104 @@
+%% -------------------------------------------------------------------
+%%
+%% riak_core: Core Riak Application
```
Comment (cmeiklejohn) [inline src/mclock.erl]: Remove this, probably.  :)

File: src/ishikawa_backend.erl [hunk 6]
```diff
@@ -47,7 +47,11 @@
 
 -record(state, {actor :: actor(),
                 vv :: timestamp(),
-                members :: [term()]}).
+                members :: [term()],
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: I think this is actually going to be `[node()]`.

File: src/ishikawa_backend.erl [hunk 7]
```diff
@@ -47,7 +47,11 @@
...
+                svv :: timestamp(),
+                rtm :: timestamp_matrix(),
+                to_be_delivered_queue :: [{timestamp(), term()}],
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: Can you be more specific about `term()` here?

File: src/ishikawa_backend.erl [hunk 8]
```diff
@@ -47,7 +47,11 @@
...
+                to_be_ack_queue :: [{term(), [term()]}]}).
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: Can you be more specific about the types here?  It's hard to tell what this is a list of...

File: src/ishikawa_backend.erl [hunk 9]
```diff
@@ -59,9 +63,9 @@ tcbcast(Message) ->
     gen_server:call(?MODULE, {tcbcast, Message}, infinity).
 
 %% Deliver a message.
--spec tcbdeliver(message(), timestamp()) -> ok.
-tcbdeliver(Message, Timestamp) ->
-    gen_server:call(?MODULE, {tcbdeliver, Message, Timestamp}, infinity).
+-spec tcbdeliver(actor(), message(), timestamp()) -> ok.
+tcbdeliver(Actor, Message, Timestamp) ->
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl:71]: What's calling this function?

File: src/ishikawa_backend.erl [hunk 10]
```diff
@@ -125,16 +142,67 @@ handle_call({tcbcast, Message}, _From, #state{actor=Actor,
     %% Transmit to membership.
     [send(Message, Peer) || Peer <- Members],
 
-    {reply, ok, State#state{vv=VClock}};
-handle_call({tcbdeliver, _Message, _Timestamp}, _From, State) ->
-    %% TODO: Implement me.
-    {reply, ok, State};
+    %% Add members to the queue of not ack messages.
+    ToBeAckQueue1 = [ToBeAckQueue | [{Message, Members}]],
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: I don't think this is doing what you want.  This is going to put the entire `ToBeAckQueue` list as the first element of `ToBeAckQueue1`.

Did you mean something like this?

`ToBeAckQueue1 = ToBeAckQueue ++ [{Message, Members}]`

File: src/ishikawa_backend.erl [hunk 11]
```diff
@@ -125,16 +142,67 @@ handle_call({tcbcast, Message}, _From, #state{actor=Actor,
...
+
+    {reply, ok, State#state{vv=VClock, to_be_ack_queue=ToBeAckQueue1}};
+handle_call({tcbdeliver, Origin, Message, Timestamp}, _From, #state{vv=VClock0,
+                                              rtm=RTM0,
+                                              to_be_delivered_queue=Queue0} = State) ->
+
+    %% Check if the message should be delivered
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: I'm not sure what this is doing.

What's the expected behavior of `check_causal_delivery`?  The way this comment is worded, it sounds like it's supposed to be a boolean?

File: src/ishikawa_backend.erl [hunk 12]
```diff
@@ -125,16 +142,67 @@ handle_call({tcbcast, Message}, _From, #state{actor=Actor,
...
+            [send(Msg, Peer) || Peer <- Members],
+
+            %% Generate message.
+            MessageAck = {tcbcastAck, Actor, VClock},
+
+            %% Send Ack back to message sender
+            send(MessageAck, From)
+    end,
+    {reply, ok, State#state{to_be_delivered_queue=Queue1}};
+handle_call({tcbcastAck, Actor, Message, VClock} = Msg, From, #state{to_be_ack_queue=QueueAck0} = State) ->
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: This would be more idiomatic if it was `tcbcast_ack`.

Comment (gyounes) [inline src/ishikawa_backend.erl]: Done.

File: src/ishikawa_backend.erl (hunk 6 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: What is node()? Where is it defined? in vclock.erl, there is vclock_node() that is a term ().

Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: `node()` is an Erlang built in, and it's what should be returned from the partisan membership library.

http://erlang.org/doc/reference_manual/distributed.html

Comment (gyounes) [inline src/ishikawa_backend.erl]: Done.

File: src/ishikawa_backend.erl (hunk 7 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: changed to all to [{actor(), message(), timestamp()}]

File: src/ishikawa_backend.erl (hunk 8 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: True. Changed to [{{actor(), timestamp()}, [node()]}]

File: src/ishikawa_backend.erl (hunk 9 above)
Comment (gyounes) [inline src/ishikawa_backend.erl:71]: What do you mean exactly? This function is called by a node that received this function and after sending it to all neighbors and receiving acks.

File: src/ishikawa_backend.erl (hunk 10 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: Thank you :)

File: src/ishikawa_backend.erl (hunk 11 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: True. I will name it causal_delivery(), or anything better you suggest. It checks and delivers an element and calls another function to loop on other elements in the queue that might become "deliverable" when a message is delivered.

File: src/mclock.erl (hunk 5 above)
Comment (gyounes) [inline src/mclock.erl]: Done.

File: src/trcb.erl [hunk 13]
```diff
@@ -23,11 +23,40 @@
...
+
+%% Check for all messages in the queue to be delivered
+%% Called upon delievery of a new message that could affect the delivery of messages in the queue
+-spec try_to_deliever([{actor(), message(), timestamp()}], {timestamp(), [{actor(), message(), timestamp()}]}) -> {timestamp(), [{actor(), message(), timestamp()}]}.
+try_to_deliever([], {VV, Queue}) -> {VV, Queue};
+try_to_deliever([{Origin, MsgVV, _Msg}=El | RQueue], {VV, Queue}=V) ->
+    case vclock:dominates(MsgVV, VV) of
+        true ->
+            NewVV = vclock:increment(Origin, VV),
+            %% TODO: Handle message,
```
Comment (cmeiklejohn) [inline src/trcb.erl]: What should be happening here?

File: src/trcb.erl [hunk 14]
```diff
@@ -23,11 +23,40 @@
...
+            try_to_deliever(Queue, {NewVV, Queue});
+        false ->
+            {VV, [Queue | [{Origin, Msg, MsgVV}]]}
+    end.
+
+%% Check for all messages in the queue to be delivered
```
Comment (cmeiklejohn) [inline src/trcb.erl]: Add `@doc`

File: src/trcb.erl [hunk 15]
```diff
@@ -23,11 +23,40 @@
...
 -callback tcbcast(message()) -> ok.
 
 %% Deliver a message.
--callback tcbdeliver(message(), timestamp()) -> ok.
+-callback tcbdeliver(actor(), message(), timestamp()) -> ok.
 
 %% Determine if a timestamp is stable.
 -callback tcbstable(timestamp()) -> {ok, boolean()}.
+
+%% check if a message should be deliver and deliver it, if not add it to the queue
```
Comment (cmeiklejohn) [inline src/trcb.erl]: Add `@doc`.

File: src/ishikawa_backend.erl [hunk 16]
```diff
@@ -101,40 +105,104 @@ init([]) ->
...
+    end,
+    {reply, ok, State#state{to_be_delivered_queue=Queue1}};
+handle_call({tcbcast_ack, Actor, Message, VClock}, From, #state{to_be_ack_queue=QueueAck0} = State) ->
+    case lists:keyfind({Actor, VClock}, 1, QueueAck0) of
+        {_, QueueMsg} ->
+            case length(QueueMsg)>0 of
+                true ->
+                    case length(QueueMsg) of
+                        1 ->
+                            QueueMsg1 = lists:delete(From, QueueMsg),
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: It seems like this could be outside of the case statement, given the behavior is exactly the same for both cases.

File: src/ishikawa_backend.erl [hunk 17]
```diff
@@ -101,40 +105,104 @@ init([]) ->
...
-handle_call(Msg, _From, State) ->
-    lager:warning("Unhandled messages: ~p", [Msg]),
-    {reply, ok, State}.
+handle_call({tcbcast, Actor, Message, VClock} = Msg, From, #state{to_be_ack_queue=Queue0, members=Members} = State) ->
+    case lists:keyfind({Actor, VClock}, 1, Queue0) of
+        {_, _} ->
+            %% Generate message.
+            MessageAck = {tcbcast_ack, Actor, Message, VClock},
+
+            Queue1 = Queue0,
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: Why is this necessary?

File: src/ishikawa_backend.erl [hunk 18]
```diff
@@ -101,40 +105,104 @@ init([]) ->
...
+
+    {reply, ok, State#state{vv=VClock, to_be_delivered_queue=Queue, svv=SVV, rtm=RTM}};
 handle_call({tcbstable, _Timestamp}, _From, State) ->
     %% TODO: Implement me.
     {reply, {ok, false}, State};
-handle_call(Msg, _From, State) ->
-    lager:warning("Unhandled messages: ~p", [Msg]),
-    {reply, ok, State}.
+handle_call({tcbcast, Actor, Message, VClock} = Msg, From, #state{to_be_ack_queue=Queue0, members=Members} = State) ->
+    case lists:keyfind({Actor, VClock}, 1, Queue0) of
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: You should be returning from the case, not exporting from the namespace.

File: src/ishikawa_backend.erl [hunk 19]
```diff
@@ -101,40 +105,104 @@ init([]) ->
...
     %% Transmit to membership.
-    [send(Message, Peer) || Peer <- Members],
+    [send(Message1, Peer) || Peer <- Members],
 
-    {reply, ok, State#state{vv=VClock}};
-handle_call({tcbdeliver, _Message, _Timestamp}, _From, State) ->
-    %% TODO: Implement me.
-    {reply, ok, State};
+    %% Add members to the queue of not ack messages.
+    ToBeAckQueue1 = ToBeAckQueue ++ [{{Actor, VClock}, Members}],
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: If for some reason the message is dropped, these need to be retried, right?  Otherwise, it will remain unacknowledged forever.

File: src/ishikawa_backend.erl (hunk 17 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: because in the state I return Queue1

File: src/trcb.erl [hunk 20]
```diff
@@ -23,11 +23,40 @@
...
+-spec causal_delivery({actor(), message(), timestamp()}, timestamp(), [{actor(), message(), timestamp()}]) -> {timestamp(), [{actor(), message(), timestamp()}]}.
+causal_delivery({Origin, Msg, MsgVV}, VV, Queue) ->
+    case vclock:dominates(MsgVV, VV) of
+        true ->
+            NewVV = vclock:increment(Origin, VV),
+            %% TODO: Handle message,
+            try_to_deliever(Queue, {NewVV, Queue});
+        false ->
+            {VV, [Queue | [{Origin, Msg, MsgVV}]]}
```
Comment (cmeiklejohn) [inline src/trcb.erl]: I think this is another case where the list concatenation is not working correctly.  

Did you mean this instead?

`Queue ++ [{Origin, Msg, MsgVV}]`

File: src/ishikawa_backend.erl (hunk 16 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: Done

File: src/trcb.erl (hunk 20 above)
Comment (gyounes) [inline src/trcb.erl]: True, done.

File: src/trcb.erl (hunk 14 above)
Comment (gyounes) [inline src/trcb.erl]: Done.

File: src/trcb.erl (hunk 13 above)
Comment (gyounes) [inline src/trcb.erl]: In the pure op we would be adding to the POLog. In general depends what do you want to do with message. Print it, etc.

File: src/ishikawa_backend.erl (hunk 18 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: Done.

File: src/ishikawa_backend.erl [hunk 21]
```diff
@@ -134,75 +136,46 @@ init([Fun]) ->
     {ok, Members} = ?PEER_SERVICE:members(),
     lager:info("Initial membership: ~p", [Members]),
 
-    {ok, #state{actor=Actor, vv=VClock, members=Members, svv=SVV, rtm=RTM, to_be_delivered_queue=ToBeDeliveredQueue, to_be_ack_queue=ToBeAckQueue, msg_handling_fun=MessageHandlingFun}}.
+    {_, TRef} = timer:send_interval(?WAIT_TIME_BEFORE_CHECK_RESEND, check_resend),
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl]: One thought: instead of `send_interval`, you could do a single timer `send_after` and then reschedule it in the `check_resend` handler: this is mainly because if processing gets to be _slower_ than the interval, you don't want them to back up, so you should only reschedule once you've finished processing.

File: src/ishikawa_backend.erl [hunk 22]
```diff
@@ -44,15 +44,17 @@
 -include("ishikawa.hrl").
 
 -define(PEER_SERVICE, partisan_peer_service).
--define(WAIT_TIME_BEFORE_RESEND, 3000).
+-define(WAIT_TIME_BEFORE_CHECK_RESEND, 5000).
+-define(WAIT_TIME_BEFORE_RESEND, 10000).
```
Comment (cmeiklejohn) [inline src/ishikawa_backend.erl:48]: What's the difference between these two?

Comment (gyounes) [inline src/ishikawa_backend.erl:48]: WAIT_TIME_BEFORE_CHECK_RESEND: the time to wait before check on all not acked messages in a process

WAIT_TIME_BEFORE__RESEND: the time to wait for resending a certain message if no acks where received
In the second one we only resend a message if time_of_check - time_previous_send > WAIT_TIME_BEFORE__RESEND

File: src/ishikawa_backend.erl (hunk 21 above)
Comment (gyounes) [inline src/ishikawa_backend.erl]: Done.

**Comment by cmeiklejohn** [comment]
Should this be merged now?
